### PR TITLE
Fix locations

### PR DESCRIPTION
**Pull request by jioter**
Fix locations

### CODE REVIEW

File: src/main/java/com/exadel/discountwebapp/location/repository/LocationRepository.java [hunk 1]
```diff
@@ -13,4 +13,6 @@
     List<Location> findAllByCountry_CountryCode(String countryCode);
 
     List<Location> findAllByCity(String city);
+
+    Location findByCountry_CountryCodeAndCityAndAddressLine(String countryCode, String city, String addressLine);
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/repository/LocationRepository.java:16]: For now this line can potentially return several Locations with the same countryCode - city - addressLine.
Let's consider adding unique composite index for (countryCode - city - addressLine) to the location

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java [hunk 2]
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
     @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
     public List<CityResponseVO> findAllCitiesByCountryCode(String countryCode) {
         List<Location> locations = locationRepository.findAllByCountry_CountryCode(countryCode);
-        List<CityResponseVO> response = new ArrayList<>();
-        locations.forEach(entity -> response.add(cityMapper.toVO(entity)));
-        return response;
+        List<CityResponseVO> allItems = new ArrayList<>();
+        locations.forEach(entity -> allItems.add(cityMapper.toVO(entity)));
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: It's better to filter all locations before mapping them with cityMapper. We can refactor it. Let's discuss on our daily call

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java [hunk 3]
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
     @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
     public List<CityResponseVO> findAllCitiesByCountryCode(String countryCode) {
         List<Location> locations = locationRepository.findAllByCountry_CountryCode(countryCode);
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: We are retrieving too much data here (which is not needed). In the nearest future we can refactor it to retrieve only what we need (custom projections?)

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java [hunk 4]
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
...
-        List<CityResponseVO> response = new ArrayList<>();
-        locations.forEach(entity -> response.add(cityMapper.toVO(entity)));
-        return response;
+        List<CityResponseVO> allItems = new ArrayList<>();
+        locations.forEach(entity -> allItems.add(cityMapper.toVO(entity)));
+        return allItems.stream()
+                .filter(el -> el.getCity() != null && !el.getCity().equals("")).distinct()
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: it's better use StringUtils.isNotBlank( ... ) from the commons lib

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java [hunk 5]
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
...
     }
 
     @Transactional(propagation = Propagation.REQUIRED)
     public LocationResponseVO create(LocationRequestVO request) {
-        return locationMapper.toVO(locationRepository.save(locationMapper.toEntity(request)));
+        Location location = locationRepository.findByCountry_CountryCodeAndCityAndAddressLine(
+                request.getCountryCode(),
+                request.getCity(),
+                request.getAddressLine());
+        return location == null ? locationMapper.toVO(locationRepository.save(locationMapper.toEntity(request))) : locationMapper.toVO(location);
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: It is more readable to write these methods in separate lines

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java [hunk 6]
```diff
@@ -9,6 +9,8 @@
 import com.exadel.discountwebapp.location.vo.city.CityResponseVO;
 import com.exadel.discountwebapp.location.vo.location.LocationRequestVO;
 import com.exadel.discountwebapp.location.vo.location.LocationResponseVO;
+import com.google.common.collect.Lists;
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: I see some Sonar issues in this class

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java [hunk 7]
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
...
-        return response;
+        List<CityResponseVO> allItems = new ArrayList<>();
+        locations.forEach(entity -> allItems.add(cityMapper.toVO(entity)));
+        return allItems.stream()
+                .filter(el -> el.getCity() != null && !el.getCity().equals("")).distinct()
+                .sorted(Comparator.comparing(CityResponseVO::getCity)).collect(Collectors.toList());
     }
 
     @Transactional(propagation = Propagation.REQUIRED)
     public LocationResponseVO create(LocationRequestVO request) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: let's rename to createIfNotExist

File: src/main/java/com/exadel/discountwebapp/location/repository/LocationRepository.java (hunk 1 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/repository/LocationRepository.java:16]: Added, but commented

Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/repository/LocationRepository.java:16]: In future we need somehow, as u said find duplicates and remove them

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java (hunk 2 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: Refactored to return list of strings

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java (hunk 3 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: Added protection, but I'm not sure if it's done right.

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java (hunk 4 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: Used, thanks

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java (hunk 5 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: I have used new line, coz if declarate new 2 variables and use them in ternary operator - then we get 500 when entity exits on third time, dunno yet, why. Returned to initial state.

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java (hunk 6 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: Fixed - removed

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java (hunk 7 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: Renamed

File: src/main/java/com/exadel/discountwebapp/location/controller/LocationController.java [hunk 8]
```diff
@@ -44,15 +43,15 @@ public LocationResponseVO getLocationById(@PathVariable Long id) {
 
     // TODO: tests for this controller
     @GetMapping("/cities")
-    public List<CityResponseVO> getAllCitiesByCountry(
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/controller/LocationController.java:47]: it seems like will break the UI after this change

File: src/main/java/com/exadel/discountwebapp/location/service/LocationService.java [hunk 9]
```diff
@@ -3,29 +3,28 @@
...
 import com.exadel.discountwebapp.location.entity.Location;
-import com.exadel.discountwebapp.location.mapper.CityMapper;
 import com.exadel.discountwebapp.location.mapper.LocationMapper;
+import com.exadel.discountwebapp.location.projections.CityView;
 import com.exadel.discountwebapp.location.repository.LocationRepository;
-import com.exadel.discountwebapp.location.vo.city.CityResponseVO;
 import com.exadel.discountwebapp.location.vo.location.LocationRequestVO;
 import com.exadel.discountwebapp.location.vo.location.LocationResponseVO;
 import lombok.RequiredArgsConstructor;
+import org.codehaus.plexus.util.StringUtils;
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: wrong library, let's use org.apache.commons.lang3.StringUtils;

Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/service/LocationService.java]: Changed to good one)

File: src/main/java/com/exadel/discountwebapp/location/controller/LocationController.java (hunk 8 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/controller/LocationController.java:47]: We just have called concerning it and they say it's okay, they will modify on front to process List of Strings

**Review by ahlukhouski2 — APPROVED** [review]

**Review by VolodymyrBushko — APPROVED** [review]

**Pull request merged**